README.md
# react-cmirror

> Codemirror Component for React.js, all Codemirror options and events are supported.

[中文](README_zh.md)|[English](README.md)

[![npm](https://img.shields.io/npm/v/react-cmirror.svg)](https://www.npmjs.com/package/react-cmirror)
[![npm](https://img.shields.io/npm/dm/react-cmirror.svg)](https://www.npmjs.com/package/react-cmirror)

## 1. Installation

```js
npm install --save react-cmirror
```

## 2. Demo & Examples

To build the examples locally, run:

```js
npm install
npm run start
```

## 3. Usage

Minimal usage:

```js
import ReactCodeMirror from 'react-cmirror';
<ReactCodeMirror value={text} />
```

## 4. Properties

- `value`:  the editor value
- `options`:  options for CodeMirror, see the [ CodeMirror Configuration](http://codemirror.net/doc/manual.html#config) for available options.
- `width`:  set width the editor, width and height can be either numbers (interpreted as pixels) or CSS units ("100%", for example). You can pass null for either of them to indicate that that dimension should not be changed.
- `height`:  set height the editor, same as width

**Note**: `width` and `height` overriding the css style of CodeMirror, if you want to change the style of CodeMirror, refer to [CodeMirror Customized Styling](http://codemirror.net/doc/manual.html#styling) for detail.

## 5. Events

CodeMirror support kinds of [Events](https://codemirror.net/doc/manual.html#events), to use these events, you need to follow the rules:

- `Uppercase` the first letter of event name, and add `on` at the beginning, as the property of editor.
- Event handle function recieve arguments as CodeMirror defined.

For example, to deal with `scroll` event, you need set property `onScroll`, and handeScroll will recieve `instance` argument :

```js
handleScroll = (instance /*CodeMirror instance*/) => {
  console.log(instance);
}

<ReactCodeMirror onScroll={this.handleScroll} />
```

## 6. Access to CodeMirror

You can get CodeMirror instance from `editor` and CodeMirror object from `codemirror` by using ref.

```js
getInstance = (instance) => {
  this.codemirror = instance.codemirror;
  this.editor = instance.editor;
}
<ReactCodeMirror ref={this.getInstance}/>
```

## 7. Language modes&Themes

**Language modes**
CodeMirror support syntax highlighting by setting language mode. These [Language Modes](https://codemirror.net/mode/) are available. By default, no modes is included, to enable this:

- import language mode from `mode` directory of codemirror
- set `mode` option  in `options` property

**Themes**
To change the color schemes of highlighting, theme option is supplied, these  [Themes](http://codemirror.net/demo/theme.html)  are avalible, to use like this:

- import theme from `theme` directory of codemirror
- set `theme` in `options` property

```js
import ReactCodeMirror from 'react-cmirror';
import 'codemirror/mode/markdown/markdown';
import 'codemirror/theme/monokai.css';

<ReactCodeMirror options={{mode: 'markdown', theme: 'monokai'}} />
```

## 8. Addon

[Addons](http://codemirror.net/doc/manual.html#addons) are used to  implement extra editor functionality, import addon after `react-cmirror`, and set options follow the instruction.

```js
import ReactCodeMirror from 'react-cmirror';
import 'codemirror/addon/display/fullscreen';
import 'codemirror/addon/display/fullscreen.css';
<ReactCodeMirror options={{fullScreen: true}} />
```

## 9. Async loading

Asynchronous loading page may cause unexpected rendering, use addon [autorefresh](http://codemirror.net/addon/display/autorefresh.js) to resolve it. Or you can call **refresh** function of CodeMirror manually after the page loaded.

## License

Copyright (c) 2017 ZiQiangWang [MIT](https://github.com/ZiQiangWang/react-cmirror/blob/master/LICENSE) Licensed.


CHANGELOG.md
# CHNAGELOG
##1.4.1(2018-06-15)

- 动态响应width和height的变化

##1.4.0(2018-06-15)

- codemirror.css的导入方式不是影响异步加载时样式错乱的原因，仍由库本身引入
- 响应codemirror的option实时变换，例如动态切换theme和mode

## 1.3.0(2018-06-13)

- 将codemirror.css提出，在外部导入

## 1.2.0(2018-06-02)

- 取消了属性class和style，添加了width和height属性，用来修改尺寸
- 修改向外暴露的codemirror相关属性，editor作为codemirror实例，codemirror为对象本身


lib/react-cm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _codemirror = require('codemirror');

var _codemirror2 = _interopRequireDefault(_codemirror);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('codemirror/lib/codemirror.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @authors ZiQiangWang
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @email   814120507@qq.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @date    2017-08-01 23:26:21
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var ReactCodeMirror = function (_Component) {
  _inherits(ReactCodeMirror, _Component);

  function ReactCodeMirror(props) {
    _classCallCheck(this, ReactCodeMirror);

    var _this = _possibleConstructorReturn(this, (ReactCodeMirror.__proto__ || Object.getPrototypeOf(ReactCodeMirror)).call(this, props));

    _this.getCodeMirror = function () {
      return _this.codemirror;
    };

    _this.getEventHandleFromProps = function () {
      var propNames = Object.keys(_this.props);
      var eventHandle = propNames.filter(function (prop) {
        var p = /^on+/;
        return p.test(prop);
      });

      var eventDict = {};
      eventHandle.forEach(function (ele) {
        eventDict[ele] = ele.replace(/^on[A-Z]/g, function (s) {
          return s.slice(2).toLowerCase();
        });
      });

      return eventDict;
    };

    _this.codemirror = undefined;
    return _this;
  }

  _createClass(ReactCodeMirror, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      // 生成codemirror实例
      this.codemirror = _codemirror2.default.fromTextArea(this.textarea, this.props.options);

      // 事件处理映射
      var eventDict = this.getEventHandleFromProps();

      Object.keys(eventDict).forEach(function (event) {
        _this2.codemirror.on(eventDict[event], _this2.props[event]);
      });

      // 初始化值
      this.codemirror.setValue(this.props.value || '');
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.value !== undefined && nextProps.value !== this.props.value) {
        this.codemirror.setValue(nextProps.value);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.codeMirror) {
        this.codeMirror.toTextArea();
      }
    }

    // 将props中所有的事件处理函数映射并保存

  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      return _react2.default.createElement(
        'div',
        { className: 'ReactCodeMirror ' + this.props.className },
        _react2.default.createElement('textarea', { ref: function ref(instance) {
            _this3.textarea = instance;
          } })
      );
    }
  }]);

  return ReactCodeMirror;
}(_react.Component);

ReactCodeMirror.defaultProps = {
  value: '',
  className: '',
  options: {}
};

ReactCodeMirror.propTypes = {
  value: _propTypes2.default.string,
  className: _propTypes2.default.string,
  options: _propTypes2.default.object
};

exports.default = ReactCodeMirror;

webpack.config.js
/**
 *
 * @authors ZiQiangWang
 * @email   814120507@qq.com
 * @date    2017-08-01 23:21:12
 */

var path = require('path');
var webpack = require('webpack');
var htmlWebpackPlugin = require('html-webpack-plugin');

// 定义路径
const PATHS = {
    root: path.resolve(__dirname),
    src: path.resolve(__dirname, 'src'),
    dist: path.resolve(__dirname, 'dist'),
    node_modules: path.resolve(__dirname, 'node_modules')
};

module.exports = {
  entry: './example/index',
  output: {
    path: PATHS.dist,
    filename: 'js/[name].bundle.js'
  },
  module: {
    rules: [
      {
        test: /\.jsx$/,
        use: ['babel-loader'],
        exclude: PATHS.node_modules,
        include: PATHS.example,
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader']
      },
    ]
  },
  plugins: [
    // html文件导出，这里的将文件导出到根目录
    new htmlWebpackPlugin({
      template: './example/index.html',
      filename: 'index.html',
      inject: 'body',
    })
  ],
  resolve: {
    extensions: ['.js','.jsx'] //后缀名自动补全
  }
}


webpack.config.prod.js
const path = require('path');
const htmlWebpackPlugin = require('html-webpack-plugin');

// 定义路径
const PATHS = {
    src: path.resolve(__dirname, 'src'),
    dist: path.resolve(__dirname, 'dist'),
    node_modules: path.resolve(__dirname, 'node_modules')
};

module.exports = {
  mode: 'production',
  entry: './src/react-cmirror',
  output: {
    filename: 'react-cmirror.min.js',
    library: "react-cmirror",
    libraryTarget: "umd"
  },
  module: {
    rules: [
      {
        test: /\.jsx$/,
        use: ['babel-loader'],
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader']
      },
    ]
  },
  resolve: {
    extensions: ['.js', '.jsx', '.css'] //后缀名自动补全
  },
  externals : {
    'react': 'react',
    'prop-types': 'prop-types',
    'codemirror': 'codemirror'
  },
}


lib/react-cmirror.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _codemirror = require('codemirror');

var _codemirror2 = _interopRequireDefault(_codemirror);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('codemirror/lib/codemirror.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @authors ZiQiangWang
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @email   814120507@qq.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @date    2017-08-01 23:26:21
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var ReactCodeMirror = function (_Component) {
  _inherits(ReactCodeMirror, _Component);

  function ReactCodeMirror(props) {
    _classCallCheck(this, ReactCodeMirror);

    var _this = _possibleConstructorReturn(this, (ReactCodeMirror.__proto__ || Object.getPrototypeOf(ReactCodeMirror)).call(this, props));

    _this.getEventHandleFromProps = function () {
      var propNames = Object.keys(_this.props);
      var eventHandle = propNames.filter(function (prop) {
        var p = /^on+/;
        return p.test(prop);
      });

      var eventDict = {};
      eventHandle.forEach(function (ele) {
        eventDict[ele] = ele.replace(/^on[A-Z]/g, function (s) {
          return s.slice(2).toLowerCase();
        });
      });

      return eventDict;
    };

    _this.codemirror = null;
    _this.editor = null;
    return _this;
  }

  _createClass(ReactCodeMirror, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      // 生成codemirror实例
      this.editor = _codemirror2.default.fromTextArea(this.textarea, this.props.options);
      // 获取CodeMirror用于获取其中的一些常量
      this.codemirror = _codemirror2.default;
      // 事件处理映射
      var eventDict = this.getEventHandleFromProps();

      Object.keys(eventDict).forEach(function (event) {
        _this2.editor.on(eventDict[event], _this2.props[event]);
      });

      var _props = this.props,
          value = _props.value,
          width = _props.width,
          height = _props.height;
      // 初始化值

      this.editor.setValue(value || '');

      // 设置尺寸
      this.editor.setSize(width || null, height || null);
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var val = this.editor.getValue();
      var next = nextProps.value;
      if (next !== undefined && next !== this.props.value && next !== val) {
        this.editor.setValue(nextProps.value);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.editor) {
        this.editor.toTextArea();
      }
    }

    // 将props中所有的事件处理函数映射并保存

  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      return _react2.default.createElement('textarea', { ref: function ref(instance) {
          _this3.textarea = instance;
        } });
    }
  }]);

  return ReactCodeMirror;
}(_react.Component);

ReactCodeMirror.defaultProps = {
  value: '',
  options: {},
  width: null,
  height: null
};

ReactCodeMirror.propTypes = {
  value: _propTypes2.default.string,
  options: _propTypes2.default.object,
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

exports.default = ReactCodeMirror;
module.exports = exports['default'];